src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -1866,6 +1872,20 @@ fn specialize<'p, 'a: 'p, 'tcx>(
                     "unexpected ctor {:?} for slice pat", constructor)
             }
         }
+
+        PatternKind::Or { ref pats } => {
+            let mut specialized_pats = vec![];
+            for pat in pats {
+                if let Some(s) = specialize(cx, &[pat], constructor, wild_patterns) {
+                    specialized_pats.extend(s);
+                }
+            }
+            if specialized_pats.is_empty() {
+                None
```

Comment:
I think this is too simplistic. We should probably only return `None` if all the subpatterns return `None` when specialised.

Reply:
Removed the specialization code, but we'll have to remember this in the follow-up PR.